sys/net/include/fib/fib.h

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2014  Martin Landsmann <Martin.Landsmann@HAW-Hamburg.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    net_fib Forwarding Information Base (FIB)
+ * @brief       FIB implementation
+ * @ingroup     net
+ *
+ * @{
+ *
+ * @file        fib.h
+ * @brief       Types and functions for FIB
+ * @author      Martin Landsmann
+ */
+
+#ifndef _FIB_H
+#define _FIB_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Reactive Routing Protocol (RRP) message content to request/reply route discovery
+ */
+typedef struct rrp_address_msg_t {
+    uint8_t *address;      /**< The pointer to the address */
+    uint8_t address_size;  /**< The address size */
+    kernel_pid_t iface_id; /**< The interface used for the next-hop reply.
+                                If it is KERNEL_PID_UNDEF, no next-hop has been found. */
+    uint32_t lifetime;     /**< The lifetime of this next-hop */
+} rrp_address_msg_t;
+
+#define FIB_MSG_RRP_SIGNAL (0x99)              /**< message type for RRP notifications         */
```

Comment:
@authmillenon I cannot get warm with the idea that the FIB or a routing protocol must know or even handle forwarding of packets. 

Lets assume we have only a single RRP on a node active.
If we pass the `pkt` to the RRP the memory for the  packet buffer is locked until the RRP sends or drops this packet. If the RRP needs some time to discover a route and If we try to send multiple packets to one or distinct destinations, the memory is filled quickly with stalled packets.  Eventually the layers 4++ must drop or queue additional packet sends.
If we now assume to have a proactive and reactive RP running at the same time, passing some `pkt` to the RRP will, due to the locked memory, hinder or even  prevent to send further packets to a known destination (next-hop entry is available in the FIB).

I (still) propose two options here:
1. upon no next-hop available, all RRPs are informed to start a route discovery and the network layer is returned with `-EHOSTUNREACH` that drops the packets in the pipeline to the specific destination (no mercy).
2. the FIB blocks until a RRP returns a next-hop or all RRPs finish their route discovery with `-EHOSTUNREACH`. 

I think option 1. is a pragmatic and deterministic behaviour. If there is no next-hop in the FIB for the destination then the packet cannot be forwarded.
If we can live with a stalled network-stack pipeline then we should go for option 2.,
since its a bit comparable with the behaviour passing the `pkt` to the RRP but does not need to handle forwarding the packet. And contrary to option 1. we are certain that the packet is forwarded if the destination is eventually reachable. 

@Lotterleben can you maybe provide some opinions here?